bindings/java/src/org/sleuthkit/datamodel/TaggingManager.java

```diff
@@ -0,0 +1,483 @@
+		}
+
+		TagSet tagSet = null;
+
+		CaseDbConnection connection = skCase.getConnection();
+		skCase.acquireSingleUserCaseWriteLock();
+		try (Statement stmt = connection.createStatement()) {
+			connection.beginTransaction();
+			// INSERT INTO tag_sets (name) VALUES('%s')
+			stmt.execute(String.format("INSERT INTO tag_sets (name) VALUES('%s')", name));
+
+			try (ResultSet resultSet = stmt.getGeneratedKeys()) {
+
+				resultSet.next();
+				int setID = resultSet.getInt(1);
+
+				List<TagName> updatedTags = new ArrayList<>();
+				if (tagNames != null) {
+					// Get all of the TagName ids they can be updated in one
+					// SQL call.
+					List<String> idList = new ArrayList<>();
+					for (TagName tagName : tagNames) {
+						idList.add(Long.toString(tagName.getId()));
+					}
+
+					stmt.executeUpdate(String.format("UPDATE tag_names SET tag_set_id = %d WHERE tag_name_id IN (%s)", setID, String.join(",", idList)));
+
+					for (TagName tagName : tagNames) {
+						updatedTags.add(new TagName(tagName.getId(),
+								tagName.getDisplayName(),
+								tagName.getDescription(),
+								tagName.getColor(),
+								tagName.getKnownStatus(),
+								setID));
+					}
+				}
+				tagSet = new TagSet(setID, name, updatedTags);
+			}
+			connection.commitTransaction();
+		} catch (SQLException ex) {
+			connection.rollbackTransaction();
+			throw new TskCoreException(String.format("Error adding tag set %s", name), ex);
+		} finally {
+			connection.close();
+			skCase.releaseSingleUserCaseWriteLock();
+		}
+
+		return tagSet;
+	}
+
+	/**
+	 * Remove a row from the tag set table. The TagNames in the TagSet will not 
+	 * be deleted, nor will any tags with the TagNames from the deleted tag set 
+	 * be deleted.
+	 *
+	 * @param name	Name of tag set to be deleted.
+	 *
+	 * @throws TskCoreException
+	 */
+	public void deletedTagSet(String name) throws TskCoreException, SQLException {
+		if (name == null || name.isEmpty()) {
+			throw new IllegalArgumentException("Error adding deleting TagSet, TagSet name must be non-empty string.");
+		}
+
+		CaseDbConnection connection = skCase.getConnection();
+		skCase.acquireSingleUserCaseWriteLock();
+		try (Statement stmt = connection.createStatement()) {
+			connection.beginTransaction();
+			String queryTemplate = "DELETE FROM tag_sets WHERE name = '%s'";
+			stmt.execute(String.format(queryTemplate, name));
+			connection.commitTransaction();
+		} catch (SQLException ex) {
+			connection.rollbackTransaction();
+			throw new TskCoreException(String.format("Error deleting tag set %s.", name), ex);
+		} finally {
+			connection.close();
+			skCase.releaseSingleUserCaseWriteLock();
+		}
+	}
+
+	/**
+	 * Add the given TagName to the TagSet.
+	 *
+	 * @param tagSet	 The tag set being added to.
+	 * @param tagName	The tag name to add to the set.
+	 *
+	 * @throws TskCoreException
+	 */
+	public void addTagNameToTagSet(TagSet tagSet, TagName tagName) throws TskCoreException {
```

Comment:
That does seem more consistent with existing code.

If we are going to return a TagSet a check should be made sure the tageName is not already in the TagSet.  If tagName is already in tagSet what the desired behavior?
1. Return the existing TagSet object
2. Make a copy of the passed in TagSet object
3. Throw an exception.

My inclination is to do 1 or 2.  I don't see a need to throw an exception unless you believe it would inform the coder that they maybe making a mistake.  Frankly I would lean towards 1 because nothing has changed.

Reply:
We can't change the TagSet passed in because we have made TagSet objects immutable, and the tag set has changed - it now has a new tag name. #3 does not apply because it is perfectly fine to add a tag name to a tag set. We have made TagSet _objects_ immutable, but we have not made tag set read only after creation. #2 is the right approach.

Reply:
Two is what I wrote on Saturday.  :-)